### PR TITLE
[feature] init firebase functions + implement addNewOrg function

### PR DESCRIPTION
**Pull request by kevinbae15**
Read the Readme.md to understand how to use the firebase functions for yourself!

Screenshot 2024-04-22 at 7 39 07 PM

Screenshot 2024-04-22 at 7 54 07 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "build": "react-scripts build",
     "test": "react-scripts test",
     "eject": "react-scripts eject",
-    "deploy-prod": "react-scripts build && firebase deploy",
+    "deploy-prod": "react-scripts build && firebase deploy --only hosting",
```
Comment (Emerson-Wenzel) [inline package.json:29]: what does "--only hosting" do?

File: functions/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+Here is an example of the `addNewOrg` function
+
+```
+curl -X POST http://127.0.0.1:5001/fir-testjuly8/us-central1/addNewOrg -H 'Content-Type: application/json' -d '{
+  "name": "Test Organization",
+  "domains": ["tufts.edu"], 
+  "exchange_location": "hello ave", 
+  "schedule": [null, null, null, null, null, null, {"start": "10:00", "end": "24:00"}]
```
Comment (Emerson-Wenzel) [inline functions/README.md:43]: let's explicitly add the dry_run argument here so new people don't get confused on why things don't get added to prod

File: functions/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+### Invoke a function
+
+You can invoke a function by hitting its url
+
+```
+http://127.0.0.1:5001/[project_id]/[region]/[function_name]
+```
+
+Here is an example of the `addNewOrg` function
```
Comment (Emerson-Wenzel) [inline functions/README.md:36]: So the motivation of having this function (as opposed to adding this in the console) is that we can do business logic validation? Additionally it gives a good example for future cloud functions? Is that correct? thanks!

File: functions/src/addNewOrg.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  const newOrgData = req.body;
+  console.log("Received request for new org: " + JSON.stringify(newOrgData));
+
+  // first validate the object and make sure no extraneous fields are added
+  for (const key in newOrgData) {
+    if (!["domains", "exchange_location", "name", "schedule", "dryRun"].includes(key)) {
+      return respondWithErrorMessage(`extraneous field defined in body: ${key}`)
+    }
+  }
```
Comment (Emerson-Wenzel) [inline functions/src/addNewOrg.js:26]: This is great.

File: functions/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+```
+curl -X POST http://127.0.0.1:5001/fir-testjuly8/us-central1/addNewOrg -H 'Content-Type: application/json' -d '{
```
Comment (Emerson-Wenzel) [inline functions/README.md:39]: Where do we define "us-central1" and "addNewOrg" Do we have to do those things in Firebase Web UI?

File: functions/README.md (hunk 2 above)
Comment (Emerson-Wenzel) [reply functions/README.md:43]: Also, seems like there is some not finished discussion about exchange_location vs schedule: 
image


For now let's land the code and we can update it later.

**Review by Emerson-Wenzel — APPROVED** [review]
Left comments to learn more. stamped

File: package.json (hunk 1 above)
Comment (kevinbae15) [reply package.json:29]: i believe it only releases the web app. ive yet to test tho

File: functions/README.md (hunk 3 above)
Comment (kevinbae15) [reply functions/README.md:36]: yea its removing the manual inserts of db and removing human error (hopefully)

File: package.json (hunk 1 above)
Comment (kevinbae15) [reply package.json:29]: the reason this is important is bc if we dont specify, it automatically deploys everything (db, functions, etc.)

File: functions/README.md (hunk 5 above)
Comment (kevinbae15) [reply functions/README.md:39]: for the region, maybe it depends on this? https://firebase.google.com/docs/functions/locations#selecting-regions_firestore-storage

addNewOrg would be defined when we deploy the function

**Review by Emerson-Wenzel — APPROVED** [review]
second stamp

**Review by Emerson-Wenzel — APPROVED** [review]
third stamp (lol)

**Pull request merged**